File: workflows/dump_simscape_series.py
#!/usr/bin/env python3
from __future__ import annotations

import json
import os
import sys
from pathlib import Path
from typing import Any, Dict, List

import click

root_dir = Path(__file__).resolve().parent.parent
if str(root_dir) not in sys.path:
    sys.path.insert(0, str(root_dir))

from shared.run_artifacts import resolve_runs_root


def _resolve_repo_root(user_repo_root: str | None) -> Path:
    if user_repo_root:
        return Path(user_repo_root).expanduser().resolve()
    cwd = Path(os.getcwd()).resolve()
    if (cwd / "models" / "simulink").exists():
        return cwd
    return root_dir


def _simlog_file_path(
    repo_root: Path,
    model_id: str,
    run_id: str,
    runs_dir_name: str | None,
) -> Path:
    model_dir = repo_root / "models" / "simulink" / model_id
    return (
        resolve_runs_root(model_dir, runs_dir_name=runs_dir_name)
        / run_id
        / "debug"
        / "simlog_segments.mat"
    )


def _resolve_model_id_from_run_id(
    repo_root: Path,
    run_id: str,
    runs_dir_name: str | None,
) -> str:
    models_root = repo_root / "models" / "simulink"
    if not models_root.exists():
        raise click.ClickException(f"Missing models root: {models_root}")
    matches: List[str] = []
    for model_dir in sorted(models_root.iterdir()):
        if not model_dir.is_dir():
            continue
        runs_dir = resolve_runs_root(model_dir, runs_dir_name=runs_dir_name)
        if not runs_dir.exists():
            continue
        candidate = runs_dir / run_id
        if candidate.is_dir():
            matches.append(model_dir.name)
    if not matches:
        raise click.ClickException(
            f"Run id not found under models/simulink/* run root '{runs_dir_name or 'runs'}': {run_id}"
        )
    if len(matches) > 1:
        raise click.ClickException(
            f"Ambiguous run id '{run_id}' found in models: {matches}. "
            "Pass --model-id to disambiguate."
        )
    return matches[0]


def _rows_to_table(rows: List[Dict[str, Any]]) -> str:
    headers = ["idx", "path", "points", "dims", "t_start", "t_end"]
    lines: List[str] = []
    lines.append(f"{headers[0]:>4} | {headers[1]:<90} | {headers[2]:>8} | {headers[3]:>4} | {headers[4]:>12} | {headers[5]:>12}")
    lines.append("-" * 150)
    for i, row in enumerate(rows, start=1):
        path = str(row.get("path", ""))
        points = int(row.get("points", 0))
        dims = int(row.get("dims", 0))
        t_start = float(row.get("t_start", float("nan")))
        t_end = float(row.get("t_end", float("nan")))
        lines.append(
            f"{i:4d} | {path:<90} | {points:8d} | {dims:4d} | {t_start:12.6g} | {t_end:12.6g}"
        )
    return "\n".join(lines)


def _rows_to_csv(rows: List[Dict[str, Any]]) -> str:
    out: List[str] = []
    header = ["path", "points", "dims", "t_start", "t_end"]
    out.append(",".join(header))
    for row in rows:
        rec = [
            str(row.get("path", "")),
            str(int(row.get("points", 0))),
            str(int(row.get("dims", 0))),
            str(float(row.get("t_start", float("nan")))),
            str(float(row.get("t_end", float("nan")))),
        ]
        sio = []
        for cell in rec:
            if "," in cell or '"' in cell:
                cell = '"' + cell.replace('"', '""') + '"'
            sio.append(cell)
        out.append(",".join(sio))
    return "\n".join(out)


@click.command()
@click.option("--run-id", required=True, type=str, help="Run id under models/simulink/<model>/<runs-dir-name>/<run-id>.")
@click.option("--model-id", default=None, type=str, help="Model folder name. If omitted, inferred from --run-id.")
@click.option("--repo-root", default=None, type=click.Path(path_type=Path), help="Repository root path.")
@click.option(
    "--runs-dir-name",
    type=str,
    default=None,
    help="Model-local run-artifact directory name (for example: runs_7161).",
)
@click.option("--use-raw-segments", is_flag=True, help="Use raw simlogSegments traversal instead of simscapeMerged.")
@click.option(
    "--format",
    "output_format",
    type=click.Choice(["table", "json", "csv"], case_sensitive=False),
    default="table",
    show_default=True,
    help="Output format.",
)
@click.option("--out", "out_path", default=None, type=click.Path(path_type=Path), help="Optional output file path.")
@click.option("--quiet", is_flag=True, help="Suppress banner output.")
def main(
    run_id: str,
    model_id: str | None,
    repo_root: Path | None,
    runs_dir_name: str | None,
    use_raw_segments: bool,
    output_format: str,
    out_path: Path | None,
    quiet: bool,
) -> None:
    repo = _resolve_repo_root(str(repo_root) if repo_root is not None else None)
    if model_id is None or not str(model_id).strip():
        model_id = _resolve_model_id_from_run_id(repo, str(run_id), runs_dir_name)
    mat_file = _simlog_file_path(repo, model_id, run_id, runs_dir_name)
    if not mat_file.exists():
        raise click.ClickException(
            f"Missing debug Simscape file: {mat_file}\n"
            "Run `env/bin/python workflows/simulate/run_pending_sims.py "
            f"{model_id} --save-simscape-mat` first."
        )

    try:
        import matlab.engine  # type: ignore
    except Exception as exc:
        raise click.ClickException(
            "MATLAB Engine for Python is not available in this environment."
        ) from exc

    eng = None
    try:
        eng = matlab.engine.start_matlab()
        eng.addpath(str(repo / "workflows"), nargout=0)
        eng.workspace["run_id_py"] = str(run_id)
        eng.workspace["model_id_py"] = str(model_id)
        eng.workspace["repo_root_py"] = str(repo)
        eng.workspace["use_raw_py"] = bool(use_raw_segments)
        eng.eval(
            "rows_tmp = dump_simscape_series(run_id_py, "
            "'ModelId', model_id_py, "
            "'RepoRoot', repo_root_py, "
            "'UseRawSegments', use_raw_py, "
            "'PrintTable', false);",
            nargout=0,
        )
        eng.eval("rows_json_tmp = jsonencode(rows_tmp);", nargout=0)
        rows_json = eng.workspace["rows_json_tmp"]
    except Exception as exc:
        raise click.ClickException(f"Failed to read Simscape series via MATLAB: {exc}") from exc
    finally:
        if eng is not None:
            try:
                eng.quit()
            except Exception:
                pass

    try:
        rows_raw = json.loads(str(rows_json))
    except Exception as exc:
        raise click.ClickException(f"Failed to parse MATLAB JSON output: {exc}") from exc

    if isinstance(rows_raw, dict):
        rows: List[Dict[str, Any]] = [rows_raw]
    elif isinstance(rows_raw, list):
        rows = [r for r in rows_raw if isinstance(r, dict)]
    else:
        rows = []

    rows.sort(key=lambda r: str(r.get("path", "")))

    if output_format == "json":
        rendered = json.dumps(rows, indent=2)
    elif output_format == "csv":
        rendered = _rows_to_csv(rows)
    else:
        rendered = _rows_to_table(rows)

    if out_path is not None:
        out_path.parent.mkdir(parents=True, exist_ok=True)
        out_path.write_text(rendered + ("\n" if not rendered.endswith("\n") else ""), encoding="utf-8")
        if not quiet:
            click.echo(f"Wrote {len(rows)} rows to {out_path}")
    else:
        if not quiet:
            click.echo(
                f"Run: {run_id} | Model: {model_id} | Rows: {len(rows)} | Source: {mat_file}"
            )
        click.echo(rendered)


if __name__ == "__main__":
    main()

File: workflows/dump_simscape_series.m
function rows = dump_simscape_series(run_id, varargin)
% dump_simscape_series Print all available Simscape series for a run.
%
% Usage:
%   dump_simscape_series("d2090a3ceea850315fee08976ebc891d")
%   dump_simscape_series("d2090a3ceea850315fee08976ebc891d", "ModelId", "BallDrop")
%   dump_simscape_series("d2090a3ceea850315fee08976ebc891d", "UseRawSegments", true)
%   dump_simscape_series("d2090a3ceea850315fee08976ebc891d", "PrintTable", false)
%
% It reads:
%   models/simulink/<ModelId>/runs/<run_id>/debug/simlog_segments.mat
%
% and prints one row per discovered series path:
%   idx | path | points | dims | t_start | t_end

    p = inputParser;
    addRequired(p, "run_id", @(x) ischar(x) || isstring(x));
    addParameter(p, "ModelId", "BallDrop", @(x) ischar(x) || isstring(x));
    addParameter(p, "RepoRoot", "", @(x) ischar(x) || isstring(x));
    addParameter(p, "UseRawSegments", false, @(x) islogical(x) || isnumeric(x));
    addParameter(p, "PrintTable", true, @(x) islogical(x) || isnumeric(x));
    parse(p, run_id, varargin{:});

    run_id = string(p.Results.run_id);
    model_id = string(p.Results.ModelId);
    use_raw_segments = logical(p.Results.UseRawSegments);
    print_table = logical(p.Results.PrintTable);
    repo_root = string(p.Results.RepoRoot);
    if strlength(repo_root) == 0
        repo_root = string(pwd);
    end

    mat_file = fullfile(char(repo_root), "models", "simulink", char(model_id), ...
        "runs", char(run_id), "debug", "simlog_segments.mat");
    if ~isfile(mat_file)
        error("dump_simscape_series:MissingFile", "Missing file: %s", mat_file);
    end

    S = load(mat_file);

    series_map = struct();
    if ~use_raw_segments && isfield(S, "simscapeMerged") && isstruct(S.simscapeMerged)
        series_map = S.simscapeMerged;
        mode = "simscapeMerged";
    else
        if ~isfield(S, "simlogSegments") || isempty(S.simlogSegments)
            error("dump_simscape_series:NoSegments", ...
                "No simlogSegments found in %s", mat_file);
        end
        for iseg = 1:numel(S.simlogSegments)
            seg_node = S.simlogSegments{iseg};
            seg_struct = i_simlog_node_to_struct(seg_node);
            if isempty(fieldnames(seg_struct))
                continue;
            end
            names = fieldnames(seg_struct);
            for i = 1:numel(names)
                key = names{i};
                if ~isfield(series_map, key)
                    series_map.(key) = seg_struct.(key);
                else
                    series_map.(key).Time = [series_map.(key).Time(:); seg_struct.(key).Time(:)];
                    series_map.(key).Data = [series_map.(key).Data; seg_struct.(key).Data];
                end
            end
        end
        mode = "simlogSegments(raw)";
    end

    names = sort(fieldnames(series_map));
    if print_table
        fprintf("Run: %s | Model: %s | Source: %s | Series: %d\n", ...
            run_id, model_id, mode, numel(names));
        fprintf("%4s | %-90s | %8s | %4s | %12s | %12s\n", ...
            "idx", "path", "points", "dims", "t_start", "t_end");
        fprintf("%s\n", repmat("-", 1, 150));
    end

    rows = struct("path", {}, "points", {}, "dims", {}, "t_start", {}, "t_end", {});
    for i = 1:numel(names)
        key = names{i};
        t = [];
        v = [];
        if isfield(series_map.(key), "Time")
            t = series_map.(key).Time;
        end
        if isfield(series_map.(key), "Data")
            v = series_map.(key).Data;
        end
        t = t(:);
        n = numel(t);
        if isempty(v)
            d = 0;
        elseif isvector(v)
            d = 1;
        else
            d = size(v, 2);
        end
        if n > 0
            t0 = t(1);
            t1 = t(end);
        else
            t0 = NaN;
            t1 = NaN;
        end

        if print_table
            fprintf("%4d | %-90s | %8d | %4d | %12.6g | %12.6g\n", i, key, n, d, t0, t1);
        end
        rows(end + 1) = struct( ... %#ok<AGROW>
            "path", key, ...
            "points", n, ...
            "dims", d, ...
            "t_start", t0, ...
            "t_end", t1);
    end
end


function sim_struct = i_simlog_node_to_struct(simlog_obj)
    sim_struct = struct();
    if isempty(simlog_obj)
        return;
    end

    stack = {simlog_obj};
    path_stack = {''};

    while ~isempty(stack)
        node = stack{1};
        node_path = path_stack{1};
        stack(1) = [];
        path_stack(1) = [];

        try
            s = node.series;
            if s.points > 0
                t = time(s);
                v = values(s);
                if isempty(node_path)
                    nm = node.id;
                else
                    nm = node_path;
                end
                fn = matlab.lang.makeValidName(strrep(nm, '.', '_'));
                sim_struct.(fn).Time = t(:);
                sim_struct.(fn).Data = v;
            end
        catch
            % Non-series nodes are expected in the tree.
        end

        f = fieldnames(node);
        for k = 1:numel(f)
            child = node.(f{k});
            if isa(child, 'simscape.logging.Node')
                if isempty(node_path)
                    new_path = f{k};
                else
                    new_path = [node_path '.' f{k}];
                end
                stack{end + 1} = child; %#ok<AGROW>
                path_stack{end + 1} = new_path; %#ok<AGROW>
            end
        end
    end
end
